src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.Negator.cs

```diff
@@ -0,0 +1,280 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings.InvertIf
+{
+    internal abstract partial class AbstractInvertIfCodeRefactoringProvider<TIfStatementSyntax, TEmbeddedStatement>
+    {
+        private const string LongLength = "LongLength";
+
+        private static readonly Dictionary<BinaryOperatorKind, BinaryOperatorKind> s_negatedBinaryMap =
```

Comment:
Since this is a generic type, we have three instance of these. would it worth it to extract out this whole logic to a toplevel type?

Reply:
Also other nested types, like StatementRange doesn't have to be generic but since it's a nested type we inherited type paraemters which makes us a generic type. would it make sense to make that a toplevel type as well?

Reply:
sorry, you'll conflict with work i did.  I extracted the 'negation' logic into a helper class so it could also be used by "use ternary expression".  I hope it's not painful to update this code.